--- a/Modules/constants.c
+++ b/Modules/constants.c
@@ -32,7 +32,8 @@
 PyObject *
 LDAPerr(int errnum)
 {
-    if (errnum >= LDAP_ERROR_MIN && errnum <= LDAP_ERROR_MAX) {
+    if (errnum >= LDAP_ERROR_MIN && errnum <= LDAP_ERROR_MAX &&
+	errobjects[errnum+LDAP_ERROR_OFFSET] != NULL) {
         PyErr_SetNone(errobjects[errnum + LDAP_ERROR_OFFSET]);
     }
     else {
@@ -72,7 +73,8 @@
         if (errnum == LDAP_NO_MEMORY)
             return PyErr_NoMemory();
 
-        if (errnum >= LDAP_ERROR_MIN && errnum <= LDAP_ERROR_MAX)
+        if (errnum >= LDAP_ERROR_MIN && errnum <= LDAP_ERROR_MAX &&
+	    errobjects[errnum+LDAP_ERROR_OFFSET] != NULL)
             errobj = errobjects[errnum + LDAP_ERROR_OFFSET];
         else
             errobj = LDAPexception_class;
