Index: os-python-ldap/Modules/options.c
===================================================================
--- os-python-ldap.orig/Modules/options.c
+++ os-python-ldap/Modules/options.c
@@ -169,9 +169,13 @@ LDAP_set_option(LDAPObject *self, int op
     case LDAP_OPT_SOCKET_BIND_ADDRESSES:
 #endif
         /* String valued options */
-        if (!PyArg_Parse(value, "s:set_option", &strval))
-            return 0;
-        ptr = strval;
+        if (value == Py_None) {
+            ptr = NULL;
+        } else {
+            if (!PyArg_Parse(value, "s:set_option", &strval))
+                return 0;
+            ptr = strval;
+        }
         break;
 
     case LDAP_OPT_TIMEOUT:
